### PR TITLE
Librimix recipes

### PR DESCRIPTION
**Pull request by JorisCos**
# About this PR 
This PR adds new recipes to the librimix dataset :

- DCCRNet
- DCUNet
- DPRNNTasNet
- DPTNet
- SuDORMRFNet
- SuDORMRFImprovedNet

All the models are trained for speech enhancement on librimix 16khz train-360.
The pretrained models have been uploaded to Hugging Face (@jonashaag can you take a look at DCUNet the performance are quite poor does it looks similar to your experiments ? )

Beside the new recipes I changed the data generation of this recipe. The local data generation i.e the csv files with the path to the audio is separated from the LIbriMix generation (it should have been in the first place).

I also introduce a new function `declipp` for declipping audio before feeding it to the ASR models. 

There are a lot of duplicates in the code as I can't make the folder local a symbolic link. This probably falls under the discussion of the recipes' design.

I realize that it's a lot of files to check... all recipes have been tested and should work though.

# Coming soon
Soon i will upload the pretrained models and results for SuDORMRFNet, SuDORMRFImprovedNet
I am also thinking about extending the evaluation with WER computation.

### CODE REVIEW

**Comment by jonashaag** [comment]
😍

**Comment by mpariente** [comment]
Oh yes, lots of review to do :joy: 
Could you highlight the points of attention you'd like in the review please?

**Comment by jonashaag** [comment]
Concering the DCU performance, could it be because of different context (only 2 s)? Can you show training loss graphs of each of the models?

Btw I'm surprised by ConvTasNet and DPRNN performance, it was much much worse for me (but I did dereverberation, not denoising)

**Comment by JorisCos** [comment]
> Concering the DCU performance, could it be because of different context (only 2 s)? Can you show training loss graphs of each of the models?
> 
> Btw I'm surprised by ConvTasNet and DPRNN performance, it was much much worse for me (but I did dereverberation, not denoising)

It might be because of the shorter context I can increase it if I reduce the batch size. I will send you the graphs.



> Oh yes, lots of review to do 😂
> Could you highlight the points of attention you'd like in the review please?

Can you check one `run.sh ` see the changes in the data generation. Also, the declipp function and one eval script. Otherwise it's pretty much the same as the first recipe

File: asteroid/dsp/declipp.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+    """
+
+    Args:
+        mix_np (numpy array): One mixture
+        est_np (numpy array): Estimates (nb_estimates, time)
+
+    """
+    for i in range(len(est_np)):
+        est_np[i] *= np.max(np.abs(mix_np)) / np.max(np.abs(est_np[i]))
+    return est_np
```
Comment (mpariente) [inline asteroid/dsp/declipp.py:14]: - This modifies the array in-place, which is a bad idea. 
- I'd use list comp here 
```python
mix_max = np.max(np.abs(mix_np)
return np.stack([est * mix_max / np.max(np.abs(est)) for est in est_np], dim=0)
```
- This needs a test.

File: egs/librimix/ConvTasNet/README.md [hunk 2]
```diff
@@ -13,3 +10,5 @@ parameters (see [here](./local/conf.yml) for more details).
 | train-360 | sep_clean |   14.7    |  15.1   |
 | train-100 | sep_noisy |   10.8    |  11.4   |
 | train-360 | sep_noisy |    12     |  12.5   |
+
+See available models [here](https://huggingface.co/JorisCos).
```
Comment (mpariente) [inline egs/librimix/ConvTasNet/README.md:14]: There are filters on the Hub: 
- You can see all Asteroid models here: https://huggingface.co/models?filter=asteroid
- All Asteroid on Libri1Mix here: https://huggingface.co/models?filter=asteroid,dataset:Libri1Mix

So two things: 
- I think we should have a LibriMix tag instead of Libri1Mix so that all LibriMix models are accessible from one query. 
 - For now, let's link to here: https://huggingface.co/models?filter=asteroid

File: asteroid/dsp/declipp.py [hunk 3]
```diff
@@ -0,0 +1,14 @@
+import numpy as np
+
+
+def declipp(mix_np, est_np):
```
Comment (mpariente) [inline asteroid/dsp/declipp.py:4]: I'm unsure about the name. Declipping is the task of restoring the signal after clipping. 
Maybe `normalize_estimates(estimates, mixture)` would make sense? The name is less pretty but more explicit.  

@jonashaag

File: asteroid/dsp/declipp.py [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+    """
+
+    Args:
+        mix_np (numpy array): One mixture
```
Comment (mpariente) [inline asteroid/dsp/declipp.py:8]: For typing to work, you need to use `np.array`.

File: asteroid/dsp/declipp.py [hunk 5]
```diff
@@ -0,0 +1,14 @@
+import numpy as np
+
+
+def declipp(mix_np, est_np):
+    """
```
Comment (mpariente) [inline asteroid/dsp/declipp.py:5]: Need a docstring explaining simply what the function does.

File: egs/librimix/ConvTasNet/local/prepare_data.sh [hunk 6]
```diff
@@ -6,17 +6,14 @@ python_path=python
...
-. generate_librimix.sh $storage_dir
-
-cd $current_dir
-$python_path local/create_local_metadata.py --librimix_dir $storage_dir/Libri$n_src"Mix"
+if [[ $n_src -le  1 ]]
+then
+  changed_n_src=2
+else
+  changed_n_src=n_src
+fi
```
Comment (mpariente) [inline egs/librimix/ConvTasNet/local/prepare_data.sh:14]: Can you explain that please?

File: egs/librimix/DCCRNet/README.md [hunk 7]
```diff
@@ -0,0 +1,20 @@
+### Results 
+
+The model was train for the `enh_single` task on Libri1Mix `train_360`.
```
Comment (mpariente) [inline egs/librimix/DCCRNet/README.md:3]:
```suggestion
The model was trained on the `enh_single` task on Libri1Mix `train_360`, at 16kHz.
```

Comment (mpariente) [reply egs/librimix/DCCRNet/README.md:3]: Same for all I guess.

File: egs/librimix/DCCRNet/README.md [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+
+On Libri1Mix min test set :
+
+``` yaml
+si_sdr: 13.329767398333798
+si_sdr_imp: 9.879986092474098
+sdr: 13.87279932997016
+sdr_imp: 10.370136530757103
+sir: Infinity
+sir_imp: NaN
```
Comment (mpariente) [inline egs/librimix/DCCRNet/README.md:15]: Can remove SIR for enhancement tasks.

File: egs/librimix/DCCRNet/eval.py [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+        return metric_list
+    try:
+        from espnet2.bin.asr_inference import Speech2Text
+        from espnet_model_zoo.downloader import ModelDownloader
+    except ModuleNotFoundError:
+        import warnings
+
+        warnings.warn("Couldn't find espnet installation. Continuing without.")
+        return metric_list
+    return metric_list + ["wer"]
```
Comment (mpariente) [inline egs/librimix/DCCRNet/eval.py:67]: This is pretty ugly but it's my fault. 
Let's keep it for now.

File: egs/librimix/DCCRNet/local/conf.yml [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+  epochs: 200
+  batch_size: 12
+  num_workers: 4
+  half_lr: yes
+  early_stop: yes
+  gradient_clipping: 5
+optim:
+  optimizer: adam
+  lr: 0.001
+  weight_decay: !!float 1e-5
```
Comment (mpariente) [inline egs/librimix/DCCRNet/local/conf.yml:25]: Missing new line

**Review by mpariente — COMMENTED** [review]
Overall, it looks good, thanks a lot. 
Few comments to address that I highlighted in the code. 

The duplicate on `local` is a disaster, but that's bad design of Asteroid recipes indeed.

File: egs/librimix/ConvTasNet/local/prepare_data.sh (hunk 6 above)
Comment (JorisCos) [reply egs/librimix/ConvTasNet/local/prepare_data.sh:14]: This is because Libri1Mix doesn't really exist it's a part of Libri2Mix. To access the right directory we do `Libri $n_src Mix` because of that the recipes with `n_src=1` doesn't work we have to change it to 2 which a bit awkward

File: asteroid/dsp/declipp.py (hunk 3 above)
Comment (JorisCos) [reply asteroid/dsp/declipp.py:4]: I will change the name but should I keep `est_np` as variable name to indicate that we only support `np.array` for now ?

Comment (mpariente) [reply asteroid/dsp/declipp.py:4]: Yes, keep the `est_np` but if you can change the order of the variables it would be cool.

File: egs/librimix/ConvTasNet/local/prepare_data.sh (hunk 6 above)
Comment (mpariente) [reply egs/librimix/ConvTasNet/local/prepare_data.sh:14]: Ok, thanks.

**Comment by mpariente** [comment]
Fix linter please

File: asteroid/dsp/normalization.py [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+
+    Args:
+        est_np (numpy array): Estimates (nb_estimates, time)
+        mix_np (numpy array): One mixture
+
+
+    """
+    mix_max = np.max(np.abs(mix_np))
+    return np.stack([est * mix_max / np.max(np.abs(est)) for est in est_np],
+                    dim=0)
```
Comment (mpariente) [inline asteroid/dsp/normalization.py:15]:
```suggestion
    return np.stack([est * mix_max / np.max(np.abs(est)) for est in est_np], dim=0)
```

**Comment by mpariente** [comment]
... When there is a commit on the browser, the tests are not ran :expressionless: 

There was again a linter problem, did you run black? 
Otherwise, it looks perfect.

**Comment by mpariente** [comment]
I spoke to fast, the tests started ^^

**Comment by JorisCos** [comment]
I forgot the test for normalization I will request your review when the recipe is completed

File: asteroid/dsp/normalization.py [hunk 12]
```diff
@@ -0,0 +1,14 @@
+import numpy as np
+
+
+def normalize_estimates(est_np, mix_np):
+    """Normalizes estimates according to the mixture maximum amplitude
+
+    Args:
+        est_np (numpy array): Estimates (nb_estimates, time)
+        mix_np (numpy array): One mixture (time)
```
Comment (mpariente) [inline asteroid/dsp/normalization.py:9]:
```suggestion
    Args:
        est_np (np.array): Estimates with shape (n_src, time).
        mix_np (np.array): One mixture with shape (time, ).
```

File: asteroid/dsp/normalization.py [hunk 13]
```diff
@@ -0,0 +1,14 @@
...
+
```
Comment (mpariente) [inline asteroid/dsp/normalization.py:10]:
```suggestion
```

**Review by mpariente — APPROVED** [review]

**Comment by mpariente** [comment]
Thanks again !
